### ISSUE TITLE
Support setup.cfg use rather than setup.py

### ISSUE THREAD

**Issue by altendky** [python-versioneer/python-versioneer#245]
I honestly don't know what form an implementation of this would take nor whether the system allows a reasonable solution presently.  But...  There is a shift towards `setup.cfg` over `setup.py`.  I am unaware of any way to define command classes in `setup.cfg`.  Maybe something relating to [in-tree backends](https://www.python.org/dev/peps/pep-0517/#in-tree-build-backends)?  Though I'm not sure how you nicely compose those for projects with their own.  And sure, by the time you include an in-tree backend, perhaps you are losing a chunk of the value of static `setup.cfg` files.

**Comment by zerothi** [comment]
To add to this, there is actually a move from setup.cfg to pyproject.toml. See #146

**Comment by altendky** [comment]
`pyproject.toml` is, for now, a place for programs to stick their configuration info.  It is not (yet, afaik, but I think there are discussions) a place to put project definition information (outside of poetry or such which is just making the project definition be poetry specific configuration).

This issue is about being able to tell setuptools to use versioneer via `setup.cfg`, not about configuring versioneer in `setup.cfg`.  versioneer already supports the latter.

**Comment by zerothi** [comment]
Yeah OK I see your distinction. It is not completely clear what the future intent with pep 621 is.

**Comment by effigies** [comment]
This sounds like a reasonable idea. I have no idea how to do it, but happy to review a PR from someone who does.

**Comment by gopackgo90** [comment]
Isn't this already supported by doing something like:
```
[metadata]
version = attr: projectname.__version__
```
?

`versioneer install` would still be needed to be changed to support creating those lines.

**Comment by effigies** [comment]
I think the correct thing would be not to use `version` at all but providing a `build-backend` that just fills in the `version` metadata for you.

**Comment by microblag** [comment]
I managed to get it working with just a pyproject.toml by doing the following:

1. Follow the no-vendor install instructions
1. delete setup.py
1. migrate the [versioneer] config in setup.cfg to [tool.versioneer] in pyproject.toml
1. add "version" to dynamic in pyproject.toml [project] section
1. set version to {attr = "myproject.__version__"} in [tool.setuptools.dynamic]

example of the pyproject.toml file i ended up with:

```
[project]
name = "myproject"
dynamic = ["version"]
requires-python = ">=3.10"
dependencies = [
    "requests",
]

[tool.setuptools.dynamic]
version = {attr = "myproject.__version__"}

[build-system]
requires = ["setuptools", "versioneer[toml]"]
build-backend = "setuptools.build_meta"

[tool.versioneer]
VCS = "git"
style = "pep440"
versionfile_source = "src/myproject/_version.py"
versionfile_build = "myproject/_version.py"
tag_prefix = "v"
parentdir_prefix = "myproject-"

[tool.setuptools.packages.find]
where = ["src"]
exclude = ["myproject.tests*"]

```

**Comment by benjaesq** [comment]
> I managed to get it working with just a pyproject.toml by doing the following:
> 
>     1. Follow the no-vendor install instructions
> 
>     2. delete setup.py
> 
>     3. migrate the [versioneer] config in setup.cfg to [tool.versioneer] in pyproject.toml
> 
>     4. add "version" to dynamic in pyproject.toml [project] section
> 
>     5. set version to {attr = "myproject.**version**"} in [tool.setuptools.dynamic]

@microblag, I wonder how this worked for you. I'm on 0.28 version and getting #369 because I don't include a setup.py nor a setup.cfg, only a pyproject.toml. Then @Callek mentioned in the same issue that "versioneer must be used with a setup.py and is not supported without one".

**Comment by effigies** [comment]
@microblag That probably works if you have very few dependencies. If you had a runtime dependency on (for example) numpy that gets imported when you `import myproject`, then you would also need `numpy` in your `build-system.requires`. But definitely a good suggestion for reasonably simple packages, or if your runtime and build-time requirements are already basically the same.

@benjaesq:

> Then @Callek mentioned in the same issue that "versioneer must be used with a setup.py and is not supported without one".

The operation of versioneer is two-fold:

1) Your `_version.py` in a source repository or editable installation calls git and gives you a new version dynamically every time you call `_version.get_versions()`. You generally call it once on import to populate `__version__`.
2) `setup.py` runs identical code inside `versioneer.py` during `setup()`, which then additionally replaces `_version.py` with a static dictionary. This is bundled in the `sdist` or `wheel` that no longer has access to the `.git` directory.

@microblag's approach basically drops mode 2 for mode 1 alone, which seems like it would work quite well. If that does work, you can probably even drop the `versioneer[toml]` requirement in your build-system, as that's only needed to replace the vendored `versioneer.py` in mode 2.

**Comment by Callek** [comment]
> > Then @Callek mentioned in the same issue that "versioneer must be used with a setup.py and is not supported without one".
> 
> The operation of versioneer is two-fold:

... Indeed thanks for the clarification here, I was a bit hasty when I said "must be used with..." since as you note here and elsewhere setuptools made some of this a bit better since the docs were last written!

**Comment by HenriqueAJNB** [comment]
> > I managed to get it working with just a pyproject.toml by doing the following:
> > ```
> > 1. Follow the no-vendor install instructions
> > 
> > 2. delete setup.py
> > 
> > 3. migrate the [versioneer] config in setup.cfg to [tool.versioneer] in pyproject.toml
> > 
> > 4. add "version" to dynamic in pyproject.toml [project] section
> > 
> > 5. set version to {attr = "myproject.**version**"} in [tool.setuptools.dynamic]
> > ```
> 
> @microblag, I wonder how this worked for you. I'm on 0.28 version and getting #369 because I don't include a setup.py nor a setup.cfg, only a pyproject.toml. Then @Callek mentioned in the same issue that "versioneer must be used with a setup.py and is not supported without one".

I'm agree. I don't have a `setup.py` in project and it's mandatory to use versioneer. Would be better to use the `pyproject.toml` only without the `setup.py`...

**Comment by claudiodsf** [comment]
> I managed to get it working with just a pyproject.toml by doing the following:
> 
> 1. Follow the no-vendor install instructions
> 2. delete setup.py
> 3. migrate the [versioneer] config in setup.cfg to [tool.versioneer] in pyproject.toml
> 4. add "version" to dynamic in pyproject.toml [project] section
> 5. set version to {attr = "myproject.**version**"} in [tool.setuptools.dynamic]

Thanks for these steps.

However, this will not create a static `_version.py` file when building the distribution with `python -m build`, resulting in a version number `0+unknown`.

For that, you still need to specify a `cmdclass` in `setup.py`, like in the following:

```python
"""
A minimal setup file for compatibility with versioneer.

All the configuration is in pyproject.toml.
"""
from setuptools import setup
import versioneer

setup(
    cmdclass=versioneer.get_cmdclass()
)
```
